cmd/kubelet/app/server.go

```diff
@@ -771,6 +731,74 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, stopCh <-chan
 	return nil
 }
 
+// buildKubeletClientConfig constructs the appropriate client config for the kubelet depending on whether
+// bootstrapping is enabled or client certificate rotation is enabled.
+func buildKubeletClientConfig(s *options.KubeletServer, nodeName types.NodeName) (*restclient.Config, func(), error) {
+	if s.RotateCertificates && utilfeature.DefaultFeatureGate.Enabled(features.RotateKubeletClientCertificate) {
+		certConfig, clientConfig, err := bootstrap.LoadClientConfig(s.KubeConfig, s.BootstrapKubeconfig, s.CertDirectory)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		newClientFn := func(current *tls.Certificate) (certificatesclient.CertificateSigningRequestInterface, error) {
+			// If we have a valid certificate, use that to fetch CSRs. Otherwise use the bootstrap
+			// credentials.
+			// XXX: When an external bootstrap source is available, it should be possible to always use that source
+			//      to retrieve new credentials.
+			config := certConfig
+			if current != nil {
```

Comment:
IIUC, if certificate manager has an expired cert and fails to rotate it, `Current()` will return that expired cert.
Can we check that `current` is still valid here and fall back to certConfig if it's not?

Reply:
Current() returns nil if the cert is expired.

Reply:
Or at least it should, that was the intended behavior a while back.  The transport rotation loop checks for the transition from valid -> invalid and uses that as the "die if this persists"

Reply:
The cached cert is only updated on successful rotation: http://go/gh/kubernetes/kubernetes/blob/3c493bc95a825f0f6eb38ef78a19811f204053ec/staging/src/k8s.io/client-go/util/certificate/certificate_manager.go#L396
And `Current()` just returns the cached cert without checking expiry http://go/gh/kubernetes/kubernetes/blob/3c493bc95a825f0f6eb38ef78a19811f204053ec/staging/src/k8s.io/client-go/util/certificate/certificate_manager.go#L204-L208

Reply:
Ah, I added the filtering to getCurrentCertificateOrBootstrap.  There's very little point in the cert manager handing me back an expired cert - can you think of a case where that makes sense?  If not, we may want to change the signature and verify callers get the right thing (callers should have to deal with nil in general).  Alternatively, we make every caller filter for a valid cert.

Reply:
ok, I agree.  I'll make that change and add tests